flake.nix

```diff
@@ -0,0 +1,33 @@
+{
+  description = "Alternative Haskell Infrastructure for Nixpkgs";
+
+  edition = 201909;
+
+  outputs = { self }: {
+    # Nix is really particular about argument names here;
+    # It requires that `overlay` and each of `overlays` have `final` and `prev` as arguments.
+    overlay = final: prev:
+    let
+      composeExtensions = f: g: self: super:
+      let
+        fApplied = f self super;
+        super' = super // fApplied;
+      in fApplied // g self super';
+    in (builtins.foldl' composeExtensions (_: _: { }) (import ./overlays))
+    final prev;
+    overlays =  {
+      release1903 = final: prev: import ./overlays/release-19.03.nix final prev;
+      wine = final: prev: import ./overlays/wine.nix final prev;
+      haskell = final: prev: import ./overlays/haskell.nix final prev;
+      bootstrap = final: prev: import ./overlays/bootstrap.nix final prev;
+      ghc = final: prev: import ./overlays/ghc.nix final prev;
+      ghc-packages = final: prev: import ./overlays/ghc-packages.nix final prev;
+      windows = final: prev: import ./overlays/windows.nix final prev;
+      armv6l-linux = final: prev: import ./overlays/armv6l-linux.nix final prev;
+      musl = final: prev: import ./overlays/musl.nix final prev;
```

Comment:
This is absolutely going to be painful to keep in sync. Can't we just `import ./overlays` and then fold over them? Or *if* we want them named. I'm happy to provide an attest in `./overlays/default.nix` assuming we can guarantee ordering. And then just reconstruct the overlays list through `attrValues`.

I'm not hung up on `self: super:` vs `final: prev:`; and I would hope that Nixpkgs args doesn't depend the argument names? Then let's just rewrite the arguments for the overlays into a more flake favourable way if that has no downsides with them being used in Nixpkgs arguments.

Reply:
Yeah, the eta-expansion of everything seems unnecessary too?

Reply:
Hmm, how can we fold over the overlays to get an attrset with meaningful names.

I'm pretty sure nixpkgs doesn't care about argument names.

Reply:
@michaelpj sadly, unless we change the actual argument names from `self: super:` to `final: prev:` this is inevitable, because flakes are really particular about this.

Reply:
> Hmm, how can we fold over the overlays to get an attrset with meaningful names.

I think Moritz was suggesting the reverse: put the attrset in `overlays/default.nix`, and then get the list back with `attrValues`.

That said, it sounds like the order in which we compose them is important, and I think `attrValues` is alphabetical based on names, so that's not ideal. I'd suggest *also* having a `combined` (`default`?) overlay in there that's the "default" composition of all the overlays in the order they are now. That way:
- If you want the default overlay just use that.
- If you want particular overlays you can pick them out.

> sadly, unless we change the actual argument names from self: super: to final: prev: this is inevitable, because flakes are really particular about this.

It cares about the *names of the parameters*? Really? Can it even get that information?

Reply:
> That said, it sounds like the order in which we compose them is important, and I think attrValues is alphabetical based on names, so that's not ideal. I'd suggest also having a combined (default?) overlay in there that's the "default" composition of all the overlays in the order they are now.

This is exactly what I've done currently -- a default `overlay` which is an ordered composition, and `overlays` as an unordered attrset if you need to pick them individually.

> It cares about the names of the parameters? Really? Can it even get that information?

Yes, it does really care about parameter names. And it can get that information because flakes are part of `nix` and not a separate library written in nix. It just inspects the AST to find out such information.

Reply:
> This is exactly what I've done currently

Sure: let's just put it in `overlays/default.nix` so it's the same interface elsewhere.

> Yes, it does really care about parameter names.

That seems pretty bad. *Especially* given that it enforces different names to the names that literally the whole of nixpkgs uses. IMO this deserves an upstream issue.

Reply:
I think the plan is to eventually replace self's and super's with final's and prev's everywhere. TBH new names make a lot more sense to me, and I think they can reduce the confusion with overlays.

However, I too dislike the way they are enforced in flakes.

Reply:
@balsoft but they are enforced in flakes, not in Nixpkgs right? so Nixpkgs should be fine with what ever names we give the overlays?

Reply:
@angerman yes, I believe this is the case.

Reply:
So let's just recode all overlays to respect the flake naming in favour of `self: super:`.

Reply:
I'm doing just this ATM :)